File: core/src/test/java/com/github/fburato/highwheelmodules/core/model/ModuleDependencyTest.java
package com.github.fburato.highwheelmodules.core.model;

import org.junit.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class ModuleDependencyTest {

  private final HWModule moduleA = HWModule.make("module A", "module A").get();
  private final HWModule moduleB = HWModule.make("module B", "module B").get();
  private final HWModule moduleBAlt = HWModule.make("module B", "module B").get();

  @Test
  public void getCountShouldReturnZeroOnNewDepedency() {
    assertThat(new ModuleDependency(moduleA, moduleB).getCount()).isEqualTo(0);
  }

  @Test
  public void getCountShouldIncreaseWhenCallingIncrementCount() {
    final ModuleDependency testee = new ModuleDependency(moduleA, moduleB);

    assertThat(testee.getCount()).isEqualTo(0);
    testee.incrementCount();
    assertThat(testee.getCount()).isEqualTo(1);
    assertThat(testee.getCount()).isEqualTo(1);
    testee.incrementCount();
    assertThat(testee.getCount()).isEqualTo(2);
  }

  @Test
  public void equalsShouldWorkOnDeepComparison() {
    final ModuleDependency testee = new ModuleDependency(moduleA, moduleB);
    final ModuleDependency alternative = new ModuleDependency(moduleA, moduleBAlt);

    assertThat(testee).isEqualTo(alternative);
  }

  @Test
  public void equalsShouldWorkOnComparisonWithNull() {
    final ModuleDependency testee = new ModuleDependency(moduleA, moduleB);

    assertThat(testee).isNotEqualTo(null);
  }

  @Test
  public void equalsShouldWorkOnComparisonWithOtherType() {
    final ModuleDependency testee = new ModuleDependency(moduleA, moduleB);

    assertThat(testee.equals(new Object())).isFalse();
  }
}

File: model/src/test/scala/com/github/fburato/highwheelmodules/model/modules/ModuleDependencyTest.scala
package com.github.fburato.highwheelmodules.model.modules

import org.mockito.scalatest.MockitoSugar
import org.scalatest.OneInstancePerTest
import org.scalatest.matchers.should.Matchers
import org.scalatest.wordspec.AnyWordSpec

class ModuleDependencyTest
    extends AnyWordSpec
    with Matchers
    with MockitoSugar
    with OneInstancePerTest {

  private val moduleA = HWModule.make("a", List("a")).get
  private val moduleB = HWModule.make("b", List("b")).get

  "count should be 0 on new dependency" in {
    ModuleDependency(moduleA, moduleB).count shouldBe 0
  }

  "count should increase on incrementCount invocation" in {
    val testee = ModuleDependency(moduleA, moduleB)

    testee.incrementCount()

    testee.count shouldBe 1
  }

  "equals should compare all elements" in {
    val testee = ModuleDependency(moduleA, moduleB)
    val alternateTestee = ModuleDependency(moduleA, HWModule.make(moduleB.name, List("b")).get)

    testee shouldEqual alternateTestee

    testee.incrementCount()

    testee should not equal alternateTestee
  }
}
